### PR TITLE
[ONNX] Add eliminate_unused_items pass

### PR DESCRIPTION
**Pull request by KsenijaS**
This PR:

- Adds eliminate_unused_items pass that removes unused inputs and initializers.
- Fixes run_embed_params function so it doesn't export unnecessary parameters.
- Removes  test_modifying_params in test_verify since it's no longer needed.

### CODE REVIEW

**Comment by KsenijaS** [comment]
@BowenBao @neginraoof please review

File: torch/csrc/jit/passes/onnx/eliminate_unused_items.cpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+void EliminateUnusedItemsONNX(Block* b, ParamMap& paramsDict) {
+  auto valsToParamsMap = buildValueToParamsMap(b, paramsDict);
+  eraseUnusedValuesFromMap(valsToParamsMap);
+  eraseUnusedBlockInputs(b);
+  buildParamsMapFromValueToParamsMap(valsToParamsMap, paramsDict);
+  return;
+}
+
+} // namespace jit
+} // namespace torch
```
[the remark this answers is not in the document]
Comment (KsenijaS) [reply torch/csrc/jit/passes/onnx/eliminate_unused_items.cpp]: it causes clang format error

**Review by houseroad — COMMENTED** [review]
A high level question, won't general lint pass and dce pass work here?

**Comment by KsenijaS** [comment]
dce pass doesn't remove unused initializers and the only unused inputs that it removes are from prim:fork node: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/dead_code_elimination.cpp#L51

File: test/onnx/test_verify.py [hunk 2]
```diff
@@ -64,21 +64,6 @@ def forward(self, x):
         with self.assertRaisesRegex(RuntimeError, "state_dict changed"):
             verify(MyModel(), x, backend)
 
-    def test_modifying_params(self):
```
Comment (BowenBao) [inline test/onnx/test_verify.py:67]: why is this test case removed?

Comment (KsenijaS) [reply test/onnx/test_verify.py:67]: this test case is meant to verify that unused initializers are not being removed when constant_fold is set to False and with this pass is no longer the case

File: test/onnx/debug_embed_params.py [hunk 3]
```diff
@@ -41,12 +41,7 @@ def run_embed_params(proto, model, input, state_dict=None, use_gpu=True):
         # sure our order is consistent with the model's order.
         # TODO: Even better: keyword arguments!
         for k in model.state_dict():
-            if k not in state_dict:
-                # Once PyTorch Module adds unnecessary parameter, the old pre-trained model does not have it.
-                # Just simply pass the new one.
-                # TODO: Please don't export unnecessary parameter.
-                parameters.append(model.state_dict()[k])
-            else:
+            if k in state_dict:
```
[the remark this answers is not in the document]
Comment (KsenijaS) [reply test/onnx/debug_embed_params.py:44]: When model has unused inputs my pass is removing them, but in debug_embed_params they are again added to the state_dict wich causes the test to fail

File: test/onnx/test_utility_funs.py [hunk 4]
```diff
@@ -675,6 +677,39 @@ def forward(self, x):
...
+                super(MyModel, self).__init__()
+                self.param = torch.nn.Parameter(torch.tensor([2.0]))
+
+            def forward(self, x):
+                y = x * x
+                self.param.data.add_(1.0)
+                return y
+
+        x = torch.tensor([1, 2])
+        # To keep the unused model parameter, need to set constant folding to False
```
Comment (BowenBao) [inline test/onnx/test_utility_funs.py:711]: remove outdated comment

**Review by BowenBao — APPROVED** [review]
minor comment inline, LGTM

File: torch/csrc/jit/passes/onnx/eliminate_unused_items.h [hunk 5]
```diff
@@ -0,0 +1,14 @@
+#pragma once
+
+#include <torch/csrc/jit/ir/ir.h>
+
+namespace torch {
+namespace jit {
+
+void EliminateUnusedItemsONNX(
```
Comment (houseroad) [inline torch/csrc/jit/passes/onnx/eliminate_unused_items.h:8]: Could you add comments to explain the major purpose of this pass. I am not clear, why dce and lint won't cover this pass without your explanation.

Comment (KsenijaS) [reply torch/csrc/jit/passes/onnx/eliminate_unused_items.h:11]: Sure, I added  the comment explaining the pass

**Comment by KsenijaS** [comment]
@houseroad flake8 and ci.pytorch.org test failure are not caused by this PR

**Comment by KsenijaS** [comment]
@houseroad clang-tidy failure is because of onnxTypeToScalarTypeMap in constant_fold pass:
https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/onnx/constant_fold.cpp#L17

Do you have any other comment or can we merge this PR?

**Comment by houseroad** [comment]
Do you want to create some constant to address the issue?

**Comment by KsenijaS** [comment]
@houseroad I added constants ::ONNX_NAMESPACE::TensorProto_DataType_* to replace numbers 1 - 12  in onnxTypeToScalarTypeMap inside constant folding pass that were causing clang-tidy failure on the CI. However clang-tidy does not have the generated onnx header so it will still fail on the CI.

**Comment by KsenijaS** [comment]
@houseroad can we merge this PR?

**Review by houseroad — COMMENTED** [review]
Can we let clang-tidy pass?

**Comment by BowenBao** [comment]
@houseroad please have a look, the clang-tidy issue is resolved. :)

**Comment by KsenijaS** [comment]
@houseroad CI is green, can we merge this PR?

**Review by houseroad — APPROVED** [review]

**Pull request closed without merging**